src/services/oracleDatapoint.ts

```diff
@@ -0,0 +1,351 @@
+    blockNumber: number;
+    blockDistance: number | null; // null if block parameter is not used
+    txHash: string;
+    txIndex: number;
+    payload: any;
+  };
+}
+
+interface Dictionary<T> {
+  [addresses: string]: T;
+}
+
+const queryMetadataOracle = `
+  SELECT *
+  FROM
+    (
+      -- newer than queried block
+      (
+        SELECT
+          txo.address as "address",
+          b.block_no AS "blockNumber",
+          CASE
+            WHEN $3::INTEGER IS NOT NULL THEN (block_no - $3)
+          END AS "blockDistance",
+          encode(tx.hash, 'hex') AS "txHash",
+          tx.block_index AS "txIndex",
+          CASE
+            WHEN $2::TEXT IS NOT NULL THEN (txm.json->$2) -- if no ticker is specified, return all tickers
+            ELSE txm.json
+          END AS "payload"
+        FROM tx_out txo
+          JOIN tx_metadata txm ON (txo.tx_id = txm.tx_id)
+          JOIN tx ON (tx.id = txm.tx_id)
+          JOIN block b ON (b.id = tx.block_id)
+        WHERE
+          txo.address = ANY ($1)
+          AND txm.key = 1968
+          AND CASE
+            WHEN $2::TEXT IS NOT NULL THEN (txm.json->$2) IS NOT NULL
+            ELSE true
+          END
+          AND CASE
+            WHEN $3::INTEGER IS NOT NULL THEN block_no >= $3
+            ELSE true
+          END
+        GROUP BY
+          txo.address,
+          tx.hash,
+          b.block_no,
+          tx.block_index,
+          txm.json
+        ORDER BY (
+            CASE
+              WHEN $3::INTEGER IS NOT NULL THEN (block_no - $3)
+            END
+          ) ASC,
+          -- if a block is specified, order by nearest first
+          (
+            CASE
+              WHEN $3::INTEGER IS NULL THEN block_no
+            END
+          ) DESC -- if a block is not specified, order by newest data
+        LIMIT CASE
+            -- default limit to 1 result, max 10
+            WHEN $4 >= 1
+            AND $4 <= 10 THEN $4
+            ELSE 1
+          END
+      )
+      UNION
+      -- older than queried block
+      (
+        SELECT
+          txo.address as "address",
+          b.block_no AS "blockNumber",
+          CASE
+            WHEN $3::INTEGER IS NOT NULL THEN ($3 - block_no)
+          END AS "blockDistance",
+          encode(tx.hash, 'hex') AS "txHash",
+          tx.block_index AS "txIndex",
+          CASE
+            WHEN $2::TEXT IS NOT NULL THEN (txm.json->$2) -- if no ticker is specified, return all tickers
+            ELSE txm.json
+          END AS "payload"
+        FROM tx_out txo
+          JOIN tx_metadata txm ON (txo.tx_id = txm.tx_id)
+          JOIN tx ON (tx.id = txm.tx_id)
+          JOIN block b ON (b.id = tx.block_id)
+        WHERE
+          txo.address = ANY ($1)
+          AND txm.key = 1968
+          AND CASE
+            WHEN $2::TEXT IS NOT NULL THEN (txm.json->$2) IS NOT NULL
+            ELSE true
+          END
+          AND CASE
+            WHEN $3::INTEGER IS NOT NULL THEN block_no < $3 IS NOT NULL
```

Comment:
@1000101 , also invalid null check here. Check all usages of `$3` plz

Reply:
Oh no, you're right. Checked and fixed.